query.go

```diff
@@ -3,46 +3,53 @@ package dht
 import (
 	"context"
 	"errors"
+
 	"github.com/libp2p/go-libp2p-core/network"
 	"github.com/libp2p/go-libp2p-core/peer"
 	pstore "github.com/libp2p/go-libp2p-core/peerstore"
 	"github.com/libp2p/go-libp2p-core/routing"
-	"github.com/libp2p/go-libp2p-kad-dht/kpeerset/peerheap"
-	"math/big"
-	"time"
 
-	"github.com/libp2p/go-libp2p-kad-dht/kpeerset"
+	"github.com/libp2p/go-libp2p-kad-dht/qpeerset"
 	kb "github.com/libp2p/go-libp2p-kbucket"
 )
 
 // ErrNoPeersQueried is returned when we failed to connect to any peers.
 var ErrNoPeersQueried = errors.New("failed to query any peers")
 
 type queryFn func(context.Context, peer.ID) ([]*peer.AddrInfo, error)
-type stopFn func(*kpeerset.SortedPeerset) bool
+type stopFn func(*qpeerset.QueryPeerset) bool
 
 // query represents a single disjoint query.
 type query struct {
 	// the query context.
 	ctx context.Context
+
 	// the cancellation function for the query context.
 	cancel context.CancelFunc
 
 	dht *IpfsDHT
 
-	// localPeers is the set of peers that need to be queried or have already been queried for this query.
-	localPeers *kpeerset.SortedPeerset
+	// seedPeers is the set of peers that seed the query
+	seedPeers []peer.ID
+
+	// queryPeers is the set of peers known by this query and their respective states.
+	queryPeers *qpeerset.QueryPeerset
+
+	// terminated is set when the first worker thread encounters the termination condition.
+	// Its role is to make sure that once termination is determined, it is sticky.
+	terminated bool
 
 	// globallyQueriedPeers is the combined set of peers queried across ALL the disjoint queries.
-	globallyQueriedPeers *peer.Set
+	globallyQueriedPeers *peer.Set // TODO: abstract this away from specifics of disjoint paths
 
 	// the function that will be used to query a single peer.
 	queryFn queryFn
 
 	// stopFn is used to determine if we should stop the WHOLE disjoint query.
-	stopFn stopFn
+	stopFn stopFn // TODO: can context cancel do the job? if, not abstract it like "cancel within the context"
```

Comment:
Can you explain what you're thinking here a bit more? Cancelling with a context is doable, but is much less friendly to do from outside this function. Instead of just calling `stopFn` at reasonable times instead every function that creates a query would have to have something like

```
go func(){
    t := time.Timer(time.Second) // Also make up a time that goes here
    select {
       case <-t.C:
          if stopFn() {
             cancelLookup()
          }
       case <-lookupCtx.Done:
          return
    }
}
```